### ISSUE TITLE
Problem regarding good practices

### ISSUE THREAD

**Issue by Astrus1234** [trevorstephens/gplearn#81]
Hello,

Thank's for this amazing package  !! I just discovered it yesterday and I want to use it for my problem.

My problem is a standard classification problem (0/1) with 180 variables both numeric and categorical (one-hot encoded). I created some features myself with field related knowledge (this is a credit problem where you're trying to predict if a person will default). I also created a range of new features by aggregation (min, max, sum, mean, var). This boost my features to 800. 

I want to use genetic programming with the help of GPlearn to create new features and improve my score (AUC). I'm currently using gradient boosting methods (XGboost and LightGbm).  

I dont have trouble understanding the Symbolic Transformer (please correct me if I'm wrong) : It compute the correlation (spearman in my case for GBM) of the new created feature with my target Y. If the correlation is superior to a certain threshold than the transform is kept in the bag of new features. I used it with the following parameters : 

After maybe an hour of training, it created 10 new features that I merged with my old features and ran an LGBM on it with a cross-validation of 5 folds (same seed as before). The result were pretty the same as before (sligthy worse). I think that I need to let gplearn run for much longer (maybe 24 hours ?) and change my configuration because I have used the default one to see how it work.

I have much more trouble understanding the Symbolic Regressor. Indeed for my problem I would have to create a new fitness to calcul the "roc_auc" on the real Y and the Y_preds. What I dont understand is the fitness in itself . The fitness function evaluate Y and Y_preds where I would thought that the best way to proceed is to evaluate for each new features the old_features + new_features on my LGBM with a cross-validation of 5 folds and return the score. What I fundamentally dont understand is that how for sofisticated models like GBM this would yield good features ?

Thank's again for this amazing package and the time spend to read my questions,
David

**Comment by trevorstephens** [comment]
Unfortunately `gplearn` does not yet implement any classification estimator. It's slated for the next release but I need to find a lot of time to get it done properly. SymbolicRegressor out of the box is strictly for regression problems.

You might be able to hack it using custom fitness measures, AND also transforming any output. All while remembering that your `y` is binary, but `y_pred` out of the regressor is a float. Perhaps a transform of `y_pred`: stardardising and then passing through a sigmoid/logit function might be a solution. 

It is something that I've been pondering but have not yet come up with a good robust solution for. Most of the literature simply jumps straight into the classification (most literature uses positive regression = 1, negative = 0) without addressing probabilities or something useful like that. A bit of a disconnect between ML and GP that is one of the things that `gplearn` aims to address.

**Comment by Astrus1234** [comment]
Hello @trevorstephens thank's for the detailled explanation :+1: 

I think I have not expressed well my thought. I dont want to estimate my function y because I have already a GBM for that. I want to engineer new features that will help my gbm to get a higher score. 

I have "hack" your SymbolicTransformer in order to do that. Here is my code : 

```
def _roc(y, y_pred, w):
    y2 = np.copy(y)
    y_pred2 = np.copy(y_pred)
    y2[y2>=1] = 1
    y2[y2<1]  = 0
    y_pred2[y_pred2>=1] = 1
    y_pred2[y_pred2<1]  = 0
    try:
        value = roc_auc_score(y2, y_pred2)
    except:
        value = 0

roc = make_fitness(_roc, greater_is_better=True)
```
I need to add a "try and except" because in some case where there is only 0 or 1 in the function to evaluate roc_auc_score from sklearn return an error.

I have a hard time figuring out what is doing the SymbolicTransformer because it's returning y_preds without any models like regression or gbm that estimate it. 
Can you explain me more in detail why the evaluation metric  take as parameter y_pred and how this value is calculated (I thought that only the regressor estimate this value) ? Furthermore why not passing as parameter the new feature generated so we can create our own evaluation ? (Like a cross-validation with a gbm and return the score for instance)

Thank's you so much,
David

**Comment by trevorstephens** [comment]
The regressor and transofrmer are actually identical. The only difference is that there is a "hall of fame" that is then whittled down to a lower number of least-correlated solutions. 

If I get your idea, I think I need to introduce the correlation metrics as first-class metrics to the regressor. That way you can generate new "smart" features and pass them into a new estimator. I've opened an issue for that in the next release.

Unfortunately, without the data to dig into, I can't easily debug why the `scikit-learn` AUC metric breaks down on your results or if your formulation is good for your problem.

As for what `y_pred` is.... It's the result of the program, taking your inputs and calculating an output. No magic. A future `SymbolicClassifier` will transform the regressor output to something like a decision function. Without knowledge of what you are doing with the regressor it's real hard to understand what the results might look like for an arbitrary `X`. Stay tuned for more metrics and a classifier :-)

**Comment by echo66** [comment]
@Astrus1234 you don't need to set a threshold when using ROC AUC as a metric. ROC AUC will evaluate how well you rank.
